### ISSUE TITLE
About function connect()

### ISSUE THREAD

**Issue by anhldbk** [mqttjs/MQTT.js#441]
I have the familiar code:

```js
var mqtt    = require('mqtt');

var options = {
  username: 'ping',
  password: '123456',
  host: '127.0.0.1',
  port: 1883,
}

var client  = mqtt.connect(options);
client.on('connect', function () {
  console.log('Connected')
});

client.on('error', function(){
  console.log('Error')
})

client.on('offline', function () {
  console.log('Offline')
});
```

My question is: How can we handle errors when invoking function `connect()`? 

Currently, here are some kind of errors & what printed out when I executed the code:

| Error | Event fired |
|-------|-------------|
|   Server is not available    |      `offline`       |
|    Invalid username or password    |     `error`        |

I wonder if we should add a callback to `connect()` to handle such failures? So we can have:

```js
var client  = mqtt.connect(options, function(err, result){
    if(err) {
        // something goes wrong
    }
})
```

@mcollina What do you think?

**Comment by anhldbk** [comment]
And if the client goes offline, I will receive multiple offline events:

```sh
$ node test2.js
Connected
Offline
Offline
Offline
Offline
...
```

Do you think we should notify client once? That means: event `offline` will be fired once when the client first disconnects from the server.

**Comment by mcollina** [comment]
offline is currently emitted even if a connection fails. I'm ok in a PR to address this.

**Comment by anhldbk** [comment]
@mcollina What's about `connect` with callbacks?

**Comment by anhldbk** [comment]
@mcollina Ping ping ping

**Comment by mcollina** [comment]
I think the current interface is ok, we should just avoid emitting out `'offline'` every time it fails to connect. Maybe we should even add a limit, just to avoid a wait-forever loop.

The reason for _not_ having a callback is that it will introduce two ways of handling errors: one at the first connection, and one afterwards. Maybe that's what we should be talking about, as it might be the underlining issue at API-level.

Also, the current state "is it connected/disconnected/connecting/waiting" is not currently well exposed.

Any other opinion?

**Comment by anhldbk** [comment]
@mcollina  you said about `two ways of handling errors`. Why don't we think it as `handling 2 types of errors`, including:
- `connection errors`: invalid username/password, offline service...
- `operation errors`:  any other one

**Comment by mcollina** [comment]
if we specify a callback, what happens if both a callback and a `'error'` handler are specified?

**Comment by RangerMauve** [comment]
What I usually do for handling connect issues, is I wrap my `connect` call in a `Promise` which resolves or rejects based on whether `connect` or `error` got emitted first.

**Comment by anhldbk** [comment]
@mcollina Seriously I'm thinking about it
@RangerMauve Would you pls show me your code?

**Comment by RangerMauve** [comment]
@anhldbk The relevant code looks like:

```
function connect(url) {
	return new Promise(function (resolve, reject) {
		var connection = mqtt.connect("ws:example.com");
		
		connection.once("connect", on_connect);
		connection.once("error", on_error);

		function on_connect() {
			connection.removeListener("error", on_error);
			resolve(connection);
		}

		function on_error(err) {
			connection.end();
			reject(err);
		}
	});
}
```

You might want to refactor it to take an instance of `connection` instead, so you can attach listeners earlier on.

**Comment by anhldbk** [comment]
@RangerMauve My approach is the same as you:

```js
class Data {
    connect(callback){
        this._client = mqtt.connect( this._conf )
        this._client.on('connect', this._onOnline.bind(this) )
        this._client.on('offline', this._onOffline.bind(this) )
        this._client.on('error', this._onError.bind(this) )
        this.once( 'status', callback )
    }

    _onOnline(){
        this._online = true
        this.emit('status') // success
    }

    _onError(){
        this.emit('status', 'status.failure.invalid_username_or_password')
    }

    _onOffline(){
        var this = this
        if( !this._online ){ // notify only once
            return
        }

        this._online = false
        this.emit('status', 'status.failure.offline')
    }            
}
```

**Comment by mcollina** [comment]
I think MQTT.js should provide this in some form, and emit offline only once.
A user should be able to query: a) if we are TCP connected and b) if we are online.

Errors should probably be more verbose too, and include a boolean to check if it was auth, at least.

**Comment by anhldbk** [comment]
@mcollina What do you mean by: `TCP connected` and `online` ?

**Comment by RangerMauve** [comment]
@anhldbk I think he meant
- When the actual socket connection (either plain TCP or Websocket, or whatever else) gets established
- When the initial handshaking packets have been exchanged

**Comment by mcollina** [comment]
Yes exactly @RangerMauve.

**Comment by mcollina** [comment]
This was fixed and released in 1.14.1.
